nni/nas/experiment/experiment.py

```diff
@@ -0,0 +1,462 @@
+        """Update the experiment manifest.
+
+        For future resume and experiment management.
+        """
+        Experiments().update_experiment(self.id, 'status', 'STOPPED')
+        Experiments().update_experiment(self.id, 'endTime', _get_current_timestamp())
+
+    def _start_engine_and_strategy(self) -> None:
+        config = self.config.canonical_copy()
+
+        _logger.debug('Creating engine and initializing strategy...')
+        self._engine = self.execution_engine_factory(config.execution_engine)
+        self._exec_model_space = self.executable_model_factory(config.model_format)
+        self.strategy.initialize(self._exec_model_space, self._engine)
+
+        if self._action == 'resume':
+            self.load_checkpoint()
+
+        if self._nni_manager_required():
+            self._send_space_to_nni_manager()
+
+        _logger.debug('Saving checkpoint before starting experiment...')
+        self.save_checkpoint()
+
+        _logger.info('Experiment initialized successfully. Starting exploration strategy...')
+
+        self.strategy.run()
+
+    def start(self, port: int = 8080, debug: bool = False, run_mode: RunMode = RunMode.Background) -> None:
+        """Start a NAS experiment.
+
+        Since NAS experiments always have strategies running in main thread,
+        :meth:`start` will not exit until the strategy finishes its run.
+
+        ``port`` and ``run_mode`` are only meaningful when :meth:`_nni_manager_required` returns true.
+
+        Parameters
+        ----------
+        port
+            Port to start NNI manager.
+        debug
+            If true, logging will be in debug mode.
+        run_mode
+            Whether to have the NNI manager in background, or foreground.
+
+        See Also
+        --------
+        nni.experiment.Experiment.start
+        """
+
+        if run_mode is not RunMode.Detach:
+            atexit.register(self.stop)
+
+        self._start_logging(debug)
+
+        if self._nni_manager_required():
+            _logger.debug('Starting NNI manager...')
+            if run_mode != RunMode.Background and self._action in ['create', 'resume']:
+                _logger.warning('Note that run_mode in NAS will only change the behavior of NNI manager. '
+                                'Strategy will still run in main thread.')
+            self._start_nni_manager(port, debug, run_mode)
+        else:
+            _logger.debug('Writing experiment to manifest...')
+            self._start_without_nni_manager()
+
+        if self._action in ['create', 'resume']:
+            self._start_engine_and_strategy()
+
+    def stop(self) -> None:
+        """Stop a NAS experiment.
+        """
+        _logger.info('Stopping experiment, please wait...')
+        atexit.unregister(self.stop)
+
+        self.save_checkpoint()
+
+        if self._nni_manager_required():
+            _logger.debug('Stopping NNI manager...')
+            self._stop_nni_manager()
```

Comment:
I have just found out the real reason that it returns before actually stop.
The stopping command is an HTTP request, and HTTP requests must be responded in several seconds before it times out.
So the only solution is to add a loop waiting for subprocess to exit.